### PR TITLE
introduce topologySpreadConstraints

### PR DESCRIPTION
**Pull request by roelandvanbatenburg**
https://github.com/Opster/opensearch-k8s-operator/issues/263

I had a go at it. But I'm not very familiar with either go or this project, so not sure if this is complete. Feel free to modify or leave suggestions.

I could not find instructions on how to test this. In `README.md` it suggests to spin up minikube and deploy to that, but it is unclear (to me) how do modify the NodePool so that is has a configured `topologySpreadConstraints`.

### CODE REVIEW

**Comment by idanl21** [comment]
Thank you @roelandvanbatenburg, but i didn't understand what you implemented here .

**Comment by roelandvanbatenburg** [comment]
> Thank you @roelandvanbatenburg, but i didn't understand what you implemented here .

I hope I implemented a way to specify [topologySpreadContraints](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/) for the open search nodes

**Comment by disaster37** [comment]
+1 We need this to avoid lost all master / data node on same time

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]
Hi @roelandvanbatenburg . Can you please add a unit test for your new functionality? 
Doesn't have to be something fancy, just to verify the contraints actually appear on the statefulset object. It will also serve as a simple regression test to make sure we do not break this feature in the future.

**Comment by roelandvanbatenburg** [comment]
> Hi @roelandvanbatenburg . Can you please add a unit test for your new functionality? Doesn't have to be something fancy, just to verify the contraints actually appear on the statefulset object. It will also serve as a simple regression test to make sure we do not break this feature in the future.

Done

File: opensearch-operator/controllers/cluster_test.go [hunk 1]
```diff
@@ -186,11 +186,15 @@ var _ = Describe("Cluster Reconciler", func() {
 			sts := &appsv1.StatefulSet{}
 			Eventually(func() error {
 				return k8sClient.Get(context.Background(), types.NamespacedName{
-					Name:      fmt.Sprintf("%s-client", OpensearchCluster.Name),
+					Name:      fmt.Sprintf("%s-master", OpensearchCluster.Name),
 					Namespace: OpensearchCluster.Namespace,
 				}, sts)
 			}, timeout, interval).Should(Succeed())
-			Expect(sts.ObjectMeta.Labels).To(HaveKeyWithValue("quux", "quut"))
+			Expect(sts.Spec.Template.Spec.TopologySpreadConstraints[0].TopologyKey).To(Equal("zone"))
```
Comment (swoehrl-mw) [inline opensearch-operator/controllers/cluster_test.go:193]: Please move this to the correct `It` block. Even that means repeating the get for the sts, but it makes it clear what is tested. Any reason why you deleted the `Expect(sts.ObjectMeta.Labels).To(HaveKeyWithValue("quux", "quut"))` line?

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]

File: opensearch-operator/controllers/cluster_test.go (hunk 1 above)
Comment (roelandvanbatenburg) [reply opensearch-operator/controllers/cluster_test.go]: Oops, I thought it was github showing a weird change, but it was me. Restored the original test and moved the test to the correct `It` block.

**Review by swoehrl-mw — APPROVED** [review]

**Pull request merged**